### DISCUSSION TITLE
TabView ActiveIndex isn't as expected

### DISCUSSION THREAD

**Discussion by DaBe1812** [primefaces/community#1320] [PrimeFaces] [answered]
I have a dynamic TabView getting it's content from a List<>.
Opening new tabs works just fine, switching between those tabs updates the ActiveIndex correctly.
Let's pretend I have 6 open tabs, so the last index ist 5. When I close tab no 2 then the last index stays 5 switching the tabs skips TabIndex 1. When I want to clear my List by closing tabs this is a problem becase the indexes in the list differ from the indexes of the tabs.
Is there a way to update the tab indexes or a workaround for this issue?

**Comment by melloware** [comment]
Please provide an executable example using the [PrimeFaces Test](https://github.com/primefaces/primefaces-test) project.  It is the only way developers can debug your problem to help.

Can you also check the Showcase "Events and Validation" example: https://www.primefaces.org/showcase/ui/panel/tabView.xhtml?jfwid=4ecd0

**Reply by DaBe1812** [reply]
HI,
never done this before. So my testcase is here:
https://github.com/DaBe1812/primefaces-problem-tabView.git

When you open the page you can open as many tabs, as you want by clicking on Details. The first issue is, that a can't tell the Index no of a closed Tab because I only can get the headline.

The next issue can be seen by closing the tabs. When you open more then two tabs and don't close the last one but any one before the last. Then you can see by switching tabs that the indexes of the closed tabs are missing. I want to remove the tabs from the openTabObjects, best way would be by index but with the wrong numbers the index in my list differs from the tabIndex.
And you can see by switching through the tabs, the last ones aren't highlighted anymore when you switch to them.

**Reply by melloware** [reply]
OK a few small observations and questions.

1. Why are you using JSP `c:forEach` instead of using the DataModel approach listed here: https://www.primefaces.org/showcase/ui/panel/tabView.xhtml  The DataModel is meant for dynamic tabs you seem to be trying to acheive?

2. Not sure if its related by why `<f:subView>` I have never in my career seen that tag used and it doesn't seem to apply here? The javadocs say...

> Container action for all JavaServer Faces core and custom component actions used on a nested page via "jsp:include" or any custom action that dynamically includes another page from the same web application, such as JSTL's "c:import".

**Reply by melloware** [reply]
Oh I see on issue 2 its because each tab opens a sub page.  

I think I know why.  If you open 3 tabs they are 0, 1, 2.  If you delete index 1 the new index are 0, 2.  It is because the component is not updating itself I think it is related to the JSP lifecycle `c:` components mixed with the JSF lifecycle components.  @tandraschko can probably answer this better but I know BalusC has said in the past "don't do this".

**Reply by tandraschko** [reply]
short: we have the model aproach for this, JSTL always have some drawbacks

**Reply by DaBe1812** [reply]
I saw this approach here https://stackoverflow.com/questions/20111010/jsf-dynamically-include-src-in-uiinclude-src-bean-pagepath 
but how would I implement a dynamic approach? 
In my project there will be about 15 different subpages and each has around 100 to 300 lines of code so I don't want to put them all in one file.
Would it be possible to try something with rendering?

**Reply by tandraschko** [reply]
why not the DataModel example here with just a simple ui:include? http://www.primefaces.org/showcase/ui/panel/tabView.xhtml?jfwid=cd5cd

**Reply by DaBe1812** [reply]
Because the page doesn't load. Something tries to interpret the page before the bean is inited. So the page in the include is ".xhtml" and this page doesn't exist und the site crashes.

**Reply by tandraschko** [reply]
please provide a reproducer primefaces-test example project and make sure you use nothing from c:* namespace

**Reply by DaBe1812** [reply]
Here is the example that crashes: https://github.com/DaBe1812/primefaces-crash.git
I have worked around the crash. 
In line 27 I have added rendered="#{openTab.page eq 'search'}"
and in line 28 I have made a statuc include on search.xhtm.
And I have copied the complete block and altered it for album.xhtml.
This works fine so far, but it seems kind of ugly.
Problem with the TabIndex stays the same, but the highlighting of the tabs is working

**Reply by Rapster** [reply]
It's a very common mistake to mix up what it's evaluated at build time and what's evaluated at render time. In your sample, you're mixing the use of `TabView#value` (which will be evaluated at render time) and `ui:include` (a tag handler which is evaluated at build time). In a few words, whenever `ui:include` is evaluated, `TabView#var` is not available yet hence path will be wrong.

First you have to understand build/render time, BalusC always provides good explanation: https://stackoverflow.com/a/31891162/4605161 and a simple example: https://stackoverflow.com/a/58981007/4605161

There is nothing wrong using JSTL and JSF tags, BUT if you don't have a clear understanding what they are and when it happens, you might encounter a lot of issues (hard to resolve). That's why, in general, the guideline is: don't mix them up if you can

But in your scenario, I don't see how you could workaround it, besides using tabView in combination with `c:foreach`. I believe the approach you shared should be helpful (https://stackoverflow.com/a/20116188/4605161)

**Reply by Rapster** [reply]
First, `TabView` has a `activeTabIndex` property which like the name implies, is the active tab... Not necessarly the tab you're listening to. To be more specific, the tab that has a `ui-state-active` class

Second, `Tabview` also has `index` property, only set if you're using var (which you can't really use in your case as explained) but I don't see why not making it always available. Ah, it's meant only for internal purposes in case of dynamic model

Third, I guess if PF had a `TabEvent#index` it should be help out?

Overall, I think third point is the way to go

@DaBe1812 As of now, what you can do is:

```java
    public void onTabClose(TabCloseEvent<?> event) {
        TabView tv = (TabView) event.getComponent();
        int activeIndex = Integer.parseInt(event.getFacesContext().getExternalContext().getRequestParameterMap().get(tv.getClientId() + "_tabindex"));

        FacesMessage msg = new FacesMessage("Closed", "Closed Tab. Active Tab " + activeIndex);
        FacesContext.getCurrentInstance().addMessage(null, msg);
    }

```

**Reply by DaBe1812** [reply]
Your suggestions seem right. 
Your Code-Snipped solves at least one of my problems because now I can tell which tab is closed and i could NULL it's place in the List and clear it's data. Also I have to check if a tab is "dirty" o I can ask the user if he wants to save his changes.

**Reply by melloware** [reply, marked as answer]
I created https://github.com/primefaces/primefaces/issues/11553 from this issue.
